### ISSUE TITLE
The association between entity types has been severed... except that is hasn't

### ISSUE THREAD

**Issue by Hyperionza** [dotnet/efcore#21307]
In a M-M relationship, when using code-first and populating initial data in the OnModelCreating for InitialCreate, it works perfectly. When I later add new data and relationships, not removing anything, and create a new migration, it fails with the error below

The table names in question are Role, GroupRole and Group.

If I change the join table to be named RoleGroup, I do NOT get the error and the migration is created successfully.

The add-migration command appears to be placing the changes to the data in alphabetical order instead of the order in which the code is actually creating the objects and relationships.

### Error

```
PM> add-migration OrgManAdminRole
Build started...
Build succeeded.
System.InvalidOperationException: The association between entity types 'Role' and 'GroupRole' has been severed but the relationship is either marked as 'Required' or is implicitly required because the foreign key is not nullable. If the dependent/child entity should be deleted when a required relationship is severed, then setup the relationship to use cascade deletes.  Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the key values.
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.HandleConceptualNulls(Boolean sensitiveLoggingEnabled, Boolean force, Boolean isCascadeDelete)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.CascadeDelete(InternalEntityEntry entry, Boolean force, IEnumerable`1 foreignKeys)
   at Microsoft.EntityFrameworkCore.Update.Internal.UpdateAdapter.CascadeDelete(IUpdateEntry entry, IEnumerable`1 foreignKeys)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.GetDataOperations(DiffContext diffContext)+MoveNext()
   at System.Linq.Enumerable.ConcatIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Sort(IEnumerable`1 operations, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.GetDifferences(IModel source, IModel target)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The association between entity types 'Role' and 'GroupRole' has been severed but the relationship is either marked as 'Required' or is implicitly required because the foreign key is not nullable. If the dependent/child entity should be deleted when a required relationship is severed, then setup the relationship to use cascade deletes.  Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the key values.
```

### Further technical details

EF Core version: 3.1.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: NET Core 3.0
Operating system: Win 10 Pro
IDE: Visual Studio 2019 16.6.2

I was getting this on earlier versions, except I just ignored it, deleted the InitialCreate migration and recreated it - It wasnt a problem yet as it was early stages of a greenfields project.
[DAL.zip](https://github.com/dotnet/efcore/files/4795149/DAL.zip)

**Comment by ajcvickers** [comment]
Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by Hyperionza** [comment]
@ajcvickers Attached to original post.
Just run an add-migration in the Package Manager Console and you should see the error

**Comment by ajcvickers** [comment]
@Hyperionza Thanks for the code--I was able to reproduce this on 3.1.5, but the issue is already fixed in 5.0 preview 5, which is available on NuGet now.

Also, you should remove references to `Microsoft.EntityFrameworkCore.SqlServer.Design` and `Microsoft.EntityFrameworkCore.Relational.Design`. These are obsolete packages that are no longer used.

**Comment by Hyperionza** [comment]
@ajcvickers Thanks for the heads up on the *.Design packages, I'm still a bit green when it comes to EFCore at times. I've probably got others that I should remove too, I'll get there in time.

Glad the issue is fixed in 5.0, I guess an upgrade is on the cards some time in November then.
